src/wp-includes/functions.wp-webfonts.php

```diff
@@ -0,0 +1,350 @@
+ *                                 number is automatically added equal to current installed WordPress version.
+ *                                 If set to null, no version is added.
+ * @param string           $media  Optional. The media for which this stylesheet has been defined.
+ *                                 Default 'screen'. Accepts media types like 'all', 'print' and 'screen', or media queries like
+ *                                 '(orientation: portrait)' and '(max-width: 640px)'.
+ */
+function wp_enqueue_webfont( $handle, $src = '', $params = array(), $ver = false, $media = 'screen' ) {
+	$params = wp_parse_args(
+		$params,
+		array(
+			'local' => false,
+		)
+	);
+
+	if ( true === $params['local'] ) {
+		$src = wp_maybe_get_local_webfont_url( $src );
+	}
+	wp_enqueue_style( "webfont-$handle", $src, array(), $ver, $media );
+}
+
+/**
+ * Remove a previously enqueued CSS stylesheet.
+ *
+ * @see WP_Dependencies::dequeue()
+ *
+ * @since 5.9.0
+ *
+ * @param string $handle Name of the stylesheet to be removed.
+ */
+function wp_dequeue_webfont( $handle ) {
+	wp_dequeue_style( "webfont-$handle" );
+}
+
+/**
+ * Check whether a webfont's CSS stylesheet has been added to the queue.
+ *
+ * @since 5.9.0
+ *
+ * @param string $handle Name of the stylesheet.
+ * @param string $list   Optional. Status of the stylesheet to check. Default 'enqueued'.
+ *                       Accepts 'enqueued', 'registered', 'queue', 'to_do', and 'done'.
+ * @return bool Whether style is queued.
+ */
+function wp_webfont_is( $handle, $list = 'enqueued' ) {
+	return wp_style_is( "webfont-$handle", $list );
+}
+
+/**
+ * Add metadata to a CSS stylesheet.
+ *
+ * Works only if the stylesheet has already been added.
+ *
+ * Possible values for $key and $value:
+ * 'conditional' string      Comments for IE 6, lte IE 7 etc.
+ * 'rtl'         bool|string To declare an RTL stylesheet.
+ * 'suffix'      string      Optional suffix, used in combination with RTL.
+ * 'alt'         bool        For rel="alternate stylesheet".
+ * 'title'       string      For preferred/alternate stylesheets.
+ *
+ * @see WP_Dependencies::add_data()
+ *
+ * @since 5.9.0
+ *
+ * @param string $handle Name of the stylesheet.
+ * @param string $key    Name of data point for which we're storing a value.
+ *                       Accepts 'conditional', 'rtl' and 'suffix', 'alt' and 'title'.
+ * @param mixed  $value  String containing the CSS data to be added.
+ * @return bool True on success, false on failure.
+ */
+function wp_webfont_add_data( $handle, $key, $value ) {
+	return wp_style_add_data( "webfont-$handle", $key, $value );
+}
+
+/**
+ * Get the local URL which contains the styles.
+ *
+ * Fallback to the remote URL if we were unable to write the file locally.
+ *
+ * @since 5.9.0
+ *
+ * @param string $remote_url The remote URL.
+ *
+ * @return string Returns the local URL if it exists, otherwise the remote URL.
+ */
+function wp_maybe_get_local_webfont_url( $remote_url ) {
```

Comment:
IMO adding some sort of local caching like this should be proposed separately. Perhaps it's even plugin territory. This was not really mentioned in https://core.trac.wordpress.org/ticket/46370 either.

If someone wants to serve a local font they should download it themselves.

Besides:

* Not everyone would want this
* Core might not need it (at least this hasn't been needed so far)
* This can cause external HTTP requests during page load on the frontend, which is bad for performance
* This does not seem to invalidate the locale cache in any way

Reply:
That's why this PR is still a draft, with a big "WIP" on the top of its description. :+1:

Reply:
Appreciate your reply, thanks! I know it's still WIP, but wanted to provide feedback early as it seemed a bit disconnected from how I understood the ticket and proposal.

For context, it's a topic that I've previously discussed with Jono and was in favor of adding such an API, hence my interest.

> > This was not really mentioned in [core.trac.wordpress.org/ticket/46370](https://core.trac.wordpress.org/ticket/46370) either.
> 
> It was mentioned in the proposal on [make.wordpress.org/core/2020/09/26/proposal-a-font-enqueue-api-for-wp-core](https://make.wordpress.org/core/2020/09/26/proposal-a-font-enqueue-api-for-wp-core/)

My understanding from the proposal and the ticket is that it's about registering either locally available or external webfonts, not downloading remote fonts for local use (although that was a [suggestion](https://make.wordpress.org/core/2020/09/26/proposal-a-font-enqueue-api-for-wp-core/#comment-39959) in the comments, mentioning all the drawbacks).

I think these two efforts should be decoupled.

> > This can cause external HTTP requests during page load on the frontend, which is bad for performance
> 
> It can cause the 1st page-load to be slower (the request that downloads the webfonts) with the benefit of making all subsequent page loads significantly faster.

Still not great. This should be done async in the background, perhaps using cron.

Related to downloading, remember that platforms like Google Fonts serve different fonts depending on the browser.

Reply:
Thank you!

> For context, it's a topic that I've previously discussed with Jono and was in favor of adding such an API, hence my interest.

That's great!

> My understanding from the proposal and the ticket is that it's about registering either locally available or external webfonts, not downloading remote fonts for local use (although that was a suggestion in the comments, mentioning all the drawbacks).

Yeah, that was going to be the next step here: Use the `$params` to generate the `@font-face` CSS when the `$src` is a font-file and not a stylesheet.
The reason I started with allowing enqueing webfonts from platforms like google-fonts and others is what most people use - and therefore it was a good starting point. Bundling font-files in themes is more rare and that is the next step needed in this implementation :+1: 

> Still not great. This should be done async in the background, perhaps using corn.

Oh that's a great idea! Thank you, I'll work on that :heart: 

> Related to downloading, remember that platforms like Google Fonts serve different fonts depending on the browser.

Since WP no longer supports IE, we don't need woff/ttf etc, and woff2 is the preferred format for all browsers. That's why the user-agent used in the request is `Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0`
AFAIK that's all that's needed to get the right font-files... Or am I missing something? :thinking:

Reply:
> Since WP no longer supports IE, we don't need woff/ttf etc, and woff2 is the preferred format for all browsers. That's why the user-agent used in the request is `Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:73.0) Gecko/20100101 Firefox/73.0`
> AFAIK that's all that's needed to get the right font-files... Or am I missing something?

Hmm good point. Didn't realize woff2 support was so good. Then should be all good I think.

Reply:
I removed the function for now in order to focus on the API first.